pkg/route/registry/route/rest.go

```diff
@@ -19,130 +20,82 @@ import (
 // HostGeneratedAnnotationKey is the key for an annotation set to "true" if the route's host was generated
 const HostGeneratedAnnotationKey = "openshift.io/host.generated"
 
-// REST is an implementation of RESTStorage for the api server.
-type REST struct {
-	registry  Registry
-	allocator route.RouteAllocator
+type routeStrategy struct {
+	runtime.ObjectTyper
+	kapi.NameGenerator
+	route.RouteAllocator
 }
 
-// NewREST returns a RESTStorage object that will work against routes.
-func NewREST(registry Registry, allocator route.RouteAllocator) *REST {
-	return &REST{
-		registry:  registry,
-		allocator: allocator,
+// NewStrategy initializes the default logic that applies when creating and updating
+// Route objects via the REST API.
+func NewStrategy(allocator route.RouteAllocator) routeStrategy {
+	return routeStrategy{
+		kapi.Scheme,
+		kapi.SimpleNameGenerator,
+		allocator,
 	}
 }
 
-// New returns a new Route
-func (rs *REST) New() runtime.Object {
-	return &api.Route{}
+func (routeStrategy) NamespaceScoped() bool {
+	return true
 }
 
-// NewList returns a new list of Routes
-func (*REST) NewList() runtime.Object {
-	return &api.Route{}
-}
-
-// List obtains a list of Routes that match label.
-func (rs *REST) List(ctx kapi.Context, label labels.Selector, field fields.Selector) (runtime.Object, error) {
-	list, err := rs.registry.ListRoutes(ctx, label)
-	if err != nil {
-		return nil, err
+func (s routeStrategy) PrepareForCreate(obj runtime.Object) {
+	route := obj.(*api.Route)
+	if len(route.Host) == 0 && s.RouteAllocator != nil {
+		// TODO: this does not belong here, and should be removed
+		shard, err := s.RouteAllocator.AllocateRouterShard(route)
+		if err != nil {
+			// TODO: this will be changed when moved to a controller
+			util.HandleError(errors.NewInternalError(fmt.Errorf("allocation error: %v for route: %#v", err, obj)))
+			return
+		}
+		route.Host = s.RouteAllocator.GenerateHostname(route, shard)
+		if route.Annotations == nil {
+			route.Annotations = map[string]string{}
+		}
+		route.Annotations[HostGeneratedAnnotationKey] = "true"
```

Comment:
@pweil- well, it seems onerous to not allow an edit on the route object, no?  And probably sorta inconsistent as well. 

@smarterclayton  - yeah, rather than have all that allocate shard + generate name cruft in here, it makes more sense to add a ```RouteAllocator.AllocateHostName(route)``` or something to that effect in here and move this code into that new method - can do that afterwards - the TODO in here is good.